cdap-master/src/main/java/co/cask/cdap/data/tools/MetricsKafkaUpgrader.java

```diff
@@ -0,0 +1,155 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.tools;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.DatasetManagementException;
+import co.cask.cdap.data2.dataset2.lib.table.MetricsTable;
+import co.cask.cdap.data2.dataset2.lib.table.hbase.HBaseTableAdmin;
+import co.cask.cdap.data2.util.TableId;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.metrics.MetricsConstants;
+import co.cask.cdap.metrics.process.KafkaConsumerMetaTable;
+import co.cask.cdap.metrics.store.DefaultMetricDatasetFactory;
+import co.cask.cdap.proto.Id;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.twill.filesystem.LocationFactory;
+import org.apache.twill.kafka.client.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Upgrade kafka metrics meta table data
+ */
+public class MetricsKafkaUpgrader extends AbstractUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DatasetUpgrader.class);
+  private static final byte[] OFFSET_COLUMN = Bytes.toBytes("o");
+
+  private final Configuration hConf;
+  private final HBaseTableUtil hBaseTableUtil;
+  private final DatasetFramework dsFramework;
+  private final String kafkaTableName;
+  private final KafkaConsumerMetaTable kafkaMetaTableDestination;
+
+  @Inject
+  public MetricsKafkaUpgrader(CConfiguration cConf, Configuration hConf, LocationFactory locationFactory,
+                              HBaseTableUtil hBaseTableUtil, @Named("dsFramework") final DatasetFramework dsFramework) {
+    super(locationFactory);
+    this.hConf = hConf;
+    this.hBaseTableUtil = hBaseTableUtil;
+    this.dsFramework = dsFramework;
+    this.kafkaTableName =  cConf.get(MetricsConstants.ConfigKeys.KAFKA_META_TABLE,
+                                     MetricsConstants.DEFAULT_KAFKA_META_TABLE);
+    this.kafkaMetaTableDestination = new DefaultMetricDatasetFactory(cConf, dsFramework).createKafkaConsumerMeta();
```

Comment:
Will this acquire resources (e.g. open HTable connection?) Does it requires closing (it's a Closeable, right?)

Ideally resources shouldn't be acquired in constructor. Since you only use it in the upgrade method, new, user and close it there.